zoo/board_games/connect4/envs/connect4_env.py

```diff
@@ -0,0 +1,674 @@
+        Note: Currently the RuleBot can only exclude actions that would lead to losing the game within three moves. One possible improvement is to further enhance the bot's long-term planning capabilities.
+Observation Space:
+    The observation in the Connect4 environment is a dictionary with five elements, which contains key information about the current state. 
+    - observation (:obj:`array`): An array that represents information about the current state, with a shape of (3, 6, 7). 
+        The length of the first dimension is 3, which stores three two-dimensional game boards with shapes (6, 7).
+        These boards represent the positions occupied by the current player, the positions occupied by the opponent player, and the identity of the current player, respectively.
+    - action_mask (:obj:`array`): A mask for the actions, indicating which actions are executable. It is a one-dimensional array of length 7, corresponding to columns 1 to 7 of the game board. 
+        It has a value of 1 for the columns where a move can be made, and a value of 0 for other positions.
+    - board (:obj:`array`): A visual representation of the current game board, represented as a 6x7 array, in which the positions where player 1 and player 2 have placed their tokens are marked with values 1 and 2, respectively. 
+    - current_player_index (:obj:`int`): The index of the current player, with player 1 having an index of 0 and player 2 having an index of 1. 
+    - to_play (:obj:`int`): The player who needs to take an action in the current state, with a value of 1 or 2.
+Action Space:
+    A set of integers from 0 to 6 (inclusive), where the action represents which column a token should be dropped in.
+Reward Space:
+    For the ``self_play_mode``, a reward of 1 is returned at the time step when the game terminates, and a reward of 0 is returned at all other time steps.
+    For the ``play_with_bot_mode``, at the time step when the game terminates, if the bot wins, the reward is -1; if the agent wins, the reward is 1; and in all other cases, the reward is 0.
+"""
+
+import os
+import sys
+import copy
+from ditk import logging
+from typing import List
+
+import numpy as np
+import pygame
+import imageio
+from ding.envs import BaseEnv, BaseEnvTimestep
+from ding.utils import ENV_REGISTRY
+from gym import spaces
+from easydict import EasyDict
+
+# from pettingzoo.utils.agent_selector import agent_selector
+from zoo.board_games.connect4.envs.rule_bot import Connect4RuleBot
+
+from zoo.board_games.mcts_bot import MCTSBot
+
+def get_image(path):
+    from os import path as os_path
+
+    import pygame
+
+    cwd = os_path.dirname(__file__)
+    image = pygame.image.load(cwd + "/" + path)
+    sfc = pygame.Surface(image.get_size(), flags=pygame.SRCALPHA)
+    sfc.blit(image, (0, 0))
+    return sfc
+
+
+@ENV_REGISTRY.register('connect4')
+class Connect4Env(BaseEnv):
+    config = dict(
```

Comment:
类似这样的格式增加详细的逐行注释哈。
```python
    # The default_config for game 2048 env.
    config = dict(
        # (str) The name of the environment registered in the environment registry.
        env_name="game_2048",
        # (bool) Whether to save the replay of the game.
        save_replay=False,
        # (str) The format in which to save the replay. 'gif' is a popular choice.
        replay_format='gif',
        # (str) A suffix for the replay file name to distinguish it from other files.
        replay_name_suffix='eval',
        # (str or None) The directory in which to save the replay file. If None, the file is saved in the current directory.
        replay_path=None,
        # (bool) Whether to render the game in real time. Useful for debugging, but can slow down training.
        render_real_time=False,
        # (bool) Whether to scale the actions. If True, actions are divided by the action space size.
        act_scale=True,
        # (bool) Whether to use the 'channel last' format for the observation space. If False, 'channel first' format is used.
        channel_last=True,
        # (str) The type of observation to use. Options are 'raw_board', 'raw_encoded_board', and 'dict_encoded_board'.
        obs_type='dict_encoded_board',
        # (bool) Whether to normalize rewards. If True, rewards are divided by the maximum possible reward.
        reward_normalize=False,
        # (float) The factor to scale rewards by when reward normalization is used.
        reward_norm_scale=100,
        # (str) The type of reward to use. 'raw' means the raw game score. 'merged_tiles_plus_log_max_tile_num' is an alternative.
        reward_type='raw',
        # (int) The maximum tile number in the game. A game is won when this tile appears. 2**11=2048, 2**16=65536
        max_tile=int(2 ** 16),
        # (int) The number of steps to delay rewards by. If > 0, the agent only receives a reward every this many steps.
        delay_reward_step=0,
        # (float) The probability that a random agent is used instead of the learning agent.
        prob_random_agent=0.,
        # (int) The maximum number of steps in an episode.
        max_episode_steps=int(1e6),
        # (bool) Whether to collect data during the game.
        is_collect=True,
        # (bool) Whether to ignore legal actions. If True, the agent can take any action, even if it's not legal.
        ignore_legal_actions=True,
        # (bool) Whether to flatten the observation space. If True, the observation space is a 1D array instead of a 2D grid.
        need_flatten=False,
        # (int) The number of possible tiles that can appear after each move.
        num_of_possible_chance_tile=2,
        # (numpy array) The possible tiles that can appear after each move.
        possible_tiles=np.array([2, 4]),
        # (numpy array) The probabilities corresponding to each possible tile.
        tile_probabilities=np.array([0.9, 0.1]),
    )
```